scripts/bert/finetune_squad.py

```diff
@@ -0,0 +1,475 @@
+if max_seq_length <= max_query_length + 3:
+    raise ValueError('The max_seq_length (%d) must be greater than max_query_length '
+                     '(%d) + 3' % (max_seq_length, max_query_length))
+
+bert, vocab = nlp.model.get_model(
+    name=model_name,
+    dataset_name=dataset_name,
+    pretrained=not model_parameters,
+    ctx=ctx,
+    use_pooler=False,
+    use_decoder=False,
+    use_classifier=False)
+
+berttoken = nlp.data.BERTTokenizer(vocab=vocab, lower=lower)
+
+net = BertForQA(bert=bert)
+if not model_parameters:
+    net.span_classifier.initialize(init=mx.init.Normal(0.02), ctx=ctx)
+else:
+    net.load_parameters(model_parameters, ctx=ctx)
+net.hybridize(static_alloc=True)
+
+loss_function = BertForQALoss()
+loss_function.hybridize(static_alloc=True)
+
+
+def train():
+    """Training function."""
+    log.info('Loader Train data...')
+    if version_2:
+        train_data = SQuAD('train', version='2.0')
+    else:
+        train_data = SQuAD('train', version='1.1')
+    log.info('Number of records in Train data:{}'.format(len(train_data)))
+
+    train_data_transform = preprocess_dataset(train_data, SQuADTransform(
+        berttoken,
+        max_seq_length=max_seq_length,
+        doc_stride=doc_stride,
+        max_query_length=max_query_length,
+        is_training=True))
+    log.info('The number of examples after preprocessing:{}'.format(
+        len(train_data_transform)))
+
+    train_dataloader = mx.gluon.data.DataLoader(
+        train_data_transform, batch_size=batch_size,
+        batchify_fn=bert_qa_batchify_fn, num_workers=4, shuffle=True)
+
+    log.info('Start Training')
+
+    optimizer_params = {'learning_rate': lr}
+    try:
+        trainer = gluon.Trainer(net.collect_params(), optimizer,
+                                optimizer_params, update_on_kvstore=False)
+    except ValueError as e:
+        print(e)
+        warnings.warn('AdamW optimizer is not found. Please consider upgrading to '
+                      'mxnet>=1.5.0. Now the original Adam optimizer is used instead.')
+        trainer = gluon.Trainer(net.collect_params(), 'adam',
+                                optimizer_params, update_on_kvstore=False)
+
+    num_train_examples = len(train_data_transform)
+    step_size = batch_size * accumulate if accumulate else batch_size
+    num_train_steps = int(num_train_examples / step_size * epochs)
+    num_warmup_steps = int(num_train_steps * warmup_ratio)
+    step_num = 0
+
+    def set_new_lr(step_num, batch_id):
+        """set new learning rate"""
+        # set grad to zero for gradient accumulation
+        if accumulate:
+            if batch_id % accumulate == 0:
+                net.collect_params().zero_grad()
+                step_num += 1
+        else:
+            step_num += 1
+        # learning rate schedule
+        if step_num < num_warmup_steps:
+            new_lr = lr * step_num / num_warmup_steps
+        else:
+            offset = (step_num - num_warmup_steps) * lr / \
```

Comment:
@eric-haibin-lin we probably need to address this.

Reply:
I didn't realize that our implementation is inconsistent with the original repo. I suggest we add a comment here that our learning rate schedule is a bit different from the original repo, so that others are aware. We probably don't have to change it and re-train everything, since we have already reproduced the result.